ocaml/xapi/xapi_vm.ml

```diff
@@ -238,6 +238,36 @@ let assert_memory_constraints ~__context ~vm platformdata =
    same VM twice during memory calculations to determine whether a given VM can start on a particular host..
 *)
 
+let update_platform_secureboot ~__context ~self platform =
+  if List.assoc "secureboot" platform = "auto" then
+    let t = ("secureboot", string_of_bool (Db.Pool.get_uefi_certificates ~__context ~self <> "")) in
+    List.append [t] (List.remove_assoc "secureboot" platform)
+  else
+    platform
+
+let save_uefi_certificates_to_dir ~__context ~self =
+  let uefi_platform_key = !Xapi_globs.varstore_dir ^ "PK.auth" in
+  if not (Sys.file_exists uefi_platform_key) then
+    begin
+      try
+        begin
+          let _dir = Sys.readdir !Xapi_globs.varstore_dir in
+          let filename = "xapi_uefi_certificates.tar" in
+          try
+            begin
+              let file = Unix.openfile filename [Unix.O_RDWR; Unix.O_CREAT] 0o755 in
+              let contents = (Xapi_stdext_base64.Base64.decode (Db.Pool.get_uefi_certificates ~__context ~self)) in
+              Unixext.write_string_to_file filename contents;
+              Tar_unix.Archive.extract (fun _ -> !Xapi_globs.varstore_dir) file;
+              finally (fun () -> Unix.close file) (fun () -> ())
+            end
+          with _ ->
+            debug "Error writing uefi certificates to file";
+        end
+      with Sys_error _ ->
+        debug "varstored directory does not exist";
+    end
+
```

Comment:
Would be good to keep track of whether this function succeeded or not, either by making it return `true/false`, or (re)raising an exception when something fails.
For this code it is probably better to reraise the exception, e.g. `failwith "Failed to save UEFI certificates: varstored directory does not exist"`. I think this will get converted to an internal error at the API level, probably worth double checking by triggerring the failure on purpose.

Reply:
The difficulty comes when the file already exists, and so the function is a noop. Has it succeeded? It technically hasn't written out the file, but it's a positive path through the function

Reply:
Do I want to reraise the exception. What's the expected path for a non-UEFI host? Should there be a condition before the function call, or should the function be a no-op on a BIOS machine (because varstored won't be running, so won't have any dir)?

Reply:
The directory is created by the host installer when the varstored package is installed, should always be there.
If it is missing nothing will notice that these files are missing, and starting a UEFI VM would fail regardless of secureboot already, so if we want to check anything related to that then secureboot is not the place for it.

Lets drop this whole `else try` branch.